org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/pom/ReImportMavenProjectsHandler.java

```diff
@@ -0,0 +1,179 @@
+      } catch (InterruptedException e) {
+        throw new OperationCanceledException(e.getMessage());
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    throw new RuntimeException("Reimport is already in progress.");
+  }
+
+  // TODO this is temporary solution.
+  // It shouldn't block response but send back projects to which a job was submitted.
+  // Then progress should be send separately.
+  // For details see https://github.com/eclipse/eclipse.jdt.ls/issues/404
+  private static Set<String> doReImportMavenProjects(
+      List<String> projectsUri, IProgressMonitor progressMonitor) throws InterruptedException {
+    final List<IProject> projects = validateProjects(projectsUri, progressMonitor);
+    jobManager.addJobChangeListener(jobChangedListener);
+    jobsToBeFinished = new CountDownLatch(projects.size());
+    submitUpdateJobs(projects);
+    jobsToBeFinished.await(REIMPORT_TIMEOUT, TimeUnit.SECONDS);
+    jobManager.removeJobChangeListener(jobChangedListener);
+    return updatedProjectsPaths.keySet();
+  }
+
+  /**
+   * Filters invalid projects (without pom) and generates list of projects to process.
+   *
+   * @param projectsUri list of URIs to projects which should be updated
+   * @param progressMonitor progress monitor
+   * @return list of valid projects to reimport
+   */
+  private static List<IProject> validateProjects(
+      List<String> projectsUri, IProgressMonitor progressMonitor) {
+    projectsToBeUpdated.clear();
+
+    List<IProject> projectsToUpdate = new ArrayList<>(projectsUri.size());
+    IFile pomFile;
+    for (String pathToProject : projectsUri) {
+      ensureNotCancelled(progressMonitor);
+
+      pomFile = JDTUtils.findFile(pathToProject + "/pom.xml");
+      if (pomFile == null) {
+        continue;
+      }
+
+      projectsToUpdate.add(pomFile.getProject());
+      projectsToBeUpdated.put(pomFile.getProject().getName(), pathToProject);
+    }
+
+    return projectsToUpdate;
+  }
+
+  /**
+   * Submits jobs to update given maven projects.
+   *
+   * @param projects list of projects to be updated
+   */
+  private static void submitUpdateJobs(List<IProject> projects) {
+    for (IProject project : projects) {
+      projectsManager.updateProject(project);
+    }
+  }
+
+  private static void ensureNotCancelled(IProgressMonitor progressMonitor) {
+    if (progressMonitor.isCanceled()) {
+      throw new OperationCanceledException();
+    }
+  }
+
+  private static class JobChangedListener extends JobChangeAdapter {
+    private static final String UPDATE_PROJECT_JOB_NAME_PREFIX = "Update project ";
+
+    /** Listens to jobs changes events to get information about updated projects. */
+    @Override
+    public void done(IJobChangeEvent event) {
+      String jobName = event.getJob().getName();
+      if (jobName.startsWith(UPDATE_PROJECT_JOB_NAME_PREFIX)) {
+        String projectName = jobName.substring(UPDATE_PROJECT_JOB_NAME_PREFIX.length());
```

Comment:
Relying on the internals of the update project job does not sound like a viable long term strategy ;-). I understand this is a temporary hack, but I've filed https://github.com/eclipse/eclipse.jdt.ls/issues/506 so we can properly identify the jobs we're interested in.

Reply:
Thank you. I'll add comment about it.